### ISSUE TITLE
Implement NRM-F

### ISSUE THREAD

**Issue by rejasupotaro** [rejasupotaro/master-thesis#1]
## Overview

Multi-field document ranking has not been studied much and NRM-F is the first and only model trying to model complex documents.

My hypothesis will be validated on NRM-F with modifications. Therefore, implementing NRM-F is the first task.

### My Hypothesis

- Query terms interact with fields differently depending on the type (e.g. "affordable speaker system").
- There is a better way than simply concatenating feature vectors (e.g. interaction-aware factorization machine).

## High-level Design

![Screenshot 2020-05-28 at 10 07 35](https://user-images.githubusercontent.com/883148/83122128-189d7900-a0cb-11ea-8d1d-f4af50d907f2.png)

## Term Representation

> We use the character n-gram representation for the following reasons: (1) it can represent out of vocabulary terms, and (2) the number of all possible tri-grams is much lower than the term-level vocabulary size which significantly reduces the number of parameters needed to be learned by the network.

Character n-gram hashing vector: https://www.microsoft.com/en-us/research/publication/learning-deep-structured-semantic-models-for-web-search-using-clickthrough-data/

## Field Representation

![formula](https://render.githubusercontent.com/render/math?math=\Phi_{F_{i}}\left(F_{i}\right)=\Lambda_{F_{i}}\left(\Phi_{f_{i}}\left(f_{i%201}\right),%20\Phi_{f_{i}}\left(f_{i%202}\right),%20\cdots,%20\Phi_{f_{i}}\left(f_{i%20m_{i}}\right)\right))

## Feature Interactions

Query-field vector is created by applying the Hadamard product. Then, query-field vectors are concatenated.

Each Phi denotes the representation learning function for each instance. The Lambda aggregates the representation of all instances. The average pooling is used as the field aggregation function.

## Loss

> Since our neural ranking model is a pairwise learning to rank model, for each query we generate all possible <q, d1, d2> triples such that the relevance label for d1 and d2 are different with respect to q.

![formula](https://render.githubusercontent.com/render/math?math=\large%20\mathcal{L}=-\frac{1}{|T|}%20\sum_{i=1}^{|T|}%20\frac{g\left(y_{i%201}\right)}{g\left(y_{i%201}\right)+g\left(y_{i%202}\right)}%20\log%20p_{i%201}+\frac{g\left(y_{i%202}\right)}{g\left(y_{i%201}\right)+g\left(y_{i%202}\right)}%20\log%20\left(1-p_{i%201}\right))

## About Data

Statistics and characteristics of the document fields.

| Field | Type | Coverage | Specific Feature |
|:---|:---|:---|:---|
| recipe_id | single instance | 100% | short text |
| user_id | single instance | 100% | short text |
| title | single instance | 100% | short text |
| description | single instance | 26% | long text |
| ingreidents | multiple instance | 100% | keywords |
| steps | multiple instance | 100% | short texts |
| country | single instance | 100% | keyword |
| created_at | single instance | 100% | date |

## Questions

- **Loss Function:**
  - Should I follow the pairwise setting? Is it practical?
- **Negative Samples:**
  - > Using explicit judgments leads to better performance compared to random negative sampling
- **Term Representation:**
  - Character-level tri-gram => Combining different granularities
  - CNN => Average Pooling?
- **Feature Interactions:**
  - Simply concatenating them is a good idea?
- **Word Embeddings**:
  - Should it be shared across fields?
  - Should I use a pre-trained embedding?
  - How to train an embedding from structured documents?
- **Local (or discrete) Representation**:
  - > To have a fair comparison, we only consider the distributed part of the model. Note that all the listed neural models, including NRM-F, can be further enriched by using lexical matching, similar to the local part of the duet model.
  - > Matching with distributed representations complements matching with traditional local representations, and that a combination of the two is favorable.
  - ![formula](https://render.githubusercontent.com/render/math?math=f(\mathbf{Q},%20\mathbf{D})=f_{\ell}(\mathbf{Q},%20\mathbf{D})%2Bf_{d}(\mathbf{Q},%20\mathbf{D}))

## Results

model | params | val_loss | val_acc | MAP | NDCG
:--|:--|:--|:--|:--|:--
Naive | 988,737 | 0.5216 | 0.7333 | 0.5427 | 0.6499
NRM-F | 1,014,977 | 0.5277 | 0.7286 | 0.4937 | 0.6095
NRM-F Concat | **984,641** | **0.5199** | **0.7344** | **0.5529** | **0.6584**

![Screenshot 2020-06-19 at 15 44 15](https://user-images.githubusercontent.com/883148/85144953-ca037a80-b243-11ea-9d9b-2277d512d0bd.png)

## Note

- Character-level n-gram representation is popular:
  - => The vocabulary size is a big deal for general search systems
- Pointwise loss vs. pairwise loss
  - => Ranking: pointwise, re-ranking pairwise
  - => When re-ranking top 100 candidates, 100*99 = 9,900 combinations are examined
- [Intro to WARP Loss, automatic differentiation and PyTorch](https://medium.com/@gabrieltseng/intro-to-warp-loss-automatic-differentiation-and-pytorch-b6aa5083187a)

## Baseline

- Simple representation based model
- https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMRanker.html

NDCG in TensorFlow
pairwise loss
